File: adops-desktop/internal/audit/audit.go
package audit

import (
	"time"

	"adops-desktop/internal/db"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type Service struct {
	gdb *gorm.DB
}

func New(gdb *gorm.DB) *Service { return &Service{gdb: gdb} }

func (s *Service) Log(userID, action, objectType string, objectID *string, result db.AuditResult, errMsg *string, old, new_ db.JSON) {
	entry := db.AuditLog{
		ID:           uuid.NewString(),
		UserID:       &userID,
		Action:       action,
		ObjectType:   objectType,
		ObjectID:     objectID,
		Result:       result,
		ErrorMessage: errMsg,
		OldValueJSON: old,
		NewValueJSON: new_,
		CreatedAt:    time.Now(),
	}
	s.gdb.Create(&entry)
}

func (s *Service) OK(userID, action, objectType, objectID string, payload db.JSON) {
	id := objectID
	s.Log(userID, action, objectType, &id, db.AuditSuccess, nil, nil, payload)
}

func (s *Service) Fail(userID, action, objectType, objectID, errMsg string) {
	id := objectID
	msg := errMsg
	s.Log(userID, action, objectType, &id, db.AuditFailed, &msg, nil, nil)
}

func (s *Service) List(userID string, limit, offset int) ([]db.AuditLog, error) {
	var logs []db.AuditLog
	err := s.gdb.
		Where("user_id = ?", userID).
		Order("created_at desc").
		Limit(limit).Offset(offset).
		Find(&logs).Error
	return logs, err
}

File: src/lib/audit.ts
import { AuditResult } from "@prisma/client";
import { SessionUser } from "./auth";
import { prisma } from "./prisma";

export async function audit(input: {
  user?: SessionUser | null;
  userId?: string | null;
  action: string;
  objectType: string;
  objectId?: string | null;
  oldValueJson?: unknown;
  newValueJson?: unknown;
  result?: AuditResult;
  errorMessage?: string | null;
}) {
  await prisma.auditLog.create({
    data: {
      userId: input.user?.id ?? input.userId ?? null,
      action: input.action,
      objectType: input.objectType,
      objectId: input.objectId,
      oldValueJson: input.oldValueJson === undefined ? undefined : (input.oldValueJson as object),
      newValueJson: input.newValueJson === undefined ? undefined : (input.newValueJson as object),
      result: input.result ?? "SUCCESS",
      errorMessage: input.errorMessage
    }
  });
}
